### PR TITLE
[clangd] Re-land "support outgoing calls in call hierarchy"

### PR DESCRIPTION
**Pull request by HighCommander4**
Previously reviewed at https://github.com/llvm/llvm-project/pull/77556 but was backed out due to build failures.

### CODE REVIEW

**Comment by HighCommander4** [comment]
I've fixed the build failure that was reported by the majority of the buildbots (which is just a new test case using a test utility function that the patch renames).

However, there is an additional build failured reported by [this buildbot](https://github.com/llvm/llvm-project/pull/77556#issuecomment-2499665944) which only applies to builds with `-DCLANGD_ENABLE_REMOTE=ON`. There is a `SymbolIndex` implementation which is only compiled in that configuration (`clang::clangd::remote::IndexClient`), which needs to be implement the new virtual method `containedRefs()` introduced by the patch.

**Review by kadircet — CHANGES_REQUESTED** [review]
sorry for losing track on the original patch, and thanks a lot for driving this to completion!

> @kadircet perhaps you might be able to pick up this review?
>
> Or, in the absence of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:
>
> > how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?

I'd definitely prefer the one we discussed in the original review, but I don't think perfect needs to be enemy of the good, we can surely optimize data structures here going forward if needed.

I am still worried about certain deployments (chromium remote-index service, and some internal ones we have). Since the optimization relies on filtering by kind, memory usage increase might be more severe for projects with different structure. 

So if it isn't going to be too annoying to ask at this point, can we make all of this a config knob? Both handling of `callHierarchy/outgoingCalls` and population of relevant structures in `DexIndex`. Option can be turned on by default, I'd like to make sure we have a quick way out if some of those deployments start OOM'ing.

**Comment by pidgeon777** [comment]
I can't find the watch button for this pull request, so I'm just commenting: hopefully this will finally be merged 🎉.

**Comment by HighCommander4** [comment]
@kadircet Thanks for having a look. I made a few updates to the PR:

 * Implemented remote index support (https://github.com/llvm/llvm-project/pull/117673/commits/d380984cecf18b53b6e2fdd4266702dcb8ea513a)
 * Implemented a config option (https://github.com/llvm/llvm-project/pull/117673/commits/02bd8f6e342c7dfbedd18c791c9e73e3c0ccb385). I did run into some challenges/limitations here:
   * Only a user config option is respected, not a project config option. I think this limitation is necessary, because the index is a global structure rather than a per-project structure. (Concretely, in the call to the `Dex` constructor that happens via `BackgroundIndexRebuilder::maybeRebuild()`, what path would I pass to the context provider that gets it to look at an appropriate project config file?)
   * Only the `Dex` logic is conditioned on the option, not the binding of the `outgoingCalls` message. This is because the binding of the message happens fairly early, while we're processing the `initialize` request. At this point we haven't loaded config files yet, and if we try, that triggers a `publishDiagnostics` message for the config file, which confuses clients who are expecting an `initialize` response.
 * Bumped index version (https://github.com/llvm/llvm-project/pull/117673/commits/c493674c8519133e1e202194fa55564e75c5a08c)

File: clang-tools-extra/clangd/ClangdServer.cpp [hunk 1]
```diff
@@ -88,9 +90,12 @@ struct UpdateIndexCallbacks : public ParsingCallbacks {
       indexStdlib(CI, std::move(*Loc));
 
     // FIndex outlives the UpdateIndexCallbacks.
-    auto Task = [FIndex(FIndex), Path(Path.str()), Version(Version.str()),
+    auto Task = [this, FIndex(FIndex), Path(Path.str()), Version(Version.str()),
                  ASTCtx(std::move(ASTCtx)), PI(std::move(PI))]() mutable {
       trace::Span Tracer("PreambleIndexing");
+      std::optional<WithContext> WithProvidedContext;
+      if (ContextProvider)
+        WithProvidedContext.emplace(ContextProvider(""));
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/ClangdServer.cpp:98]: @kadircet Not sure if this is important: it disables the building of the `RevRefs` data structure for the dynamic index.

While we do have a `Path` available to pass to the context provider here, I'm not passing it to remain consistent with the background index codepath, which only looks at the user config as explained in my previous comment.

(Also, is it fine to capture `this` in this lambda, lifetime-wise?)

**Review by kadircet — CHANGES_REQUESTED** [review]
I am sorry for my poor choice of words around `config knob`, I was trying to imply a knob like https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ClangdServer.h#L111. I don't think the knob needs to be exposed to end-users directly, only having programmatic access is fine initially. if you choose to expose it right away, i think a command line flag is the most natural way, rather than a config field. 

I'd rather plumb it explicitly through the interfaces of file-index/file-symbol/dex (and others), i think using `context` here is somewhat abusing the functionality. if you feel like that's too much (especially considering the size of the patch), feel free to leave a FIXME at least, to plumb it properly later on.

> Only a user config option is respected, not a project config option. I think this limitation is necessary, because the index is a global structure rather than a per-project structure. (Concretely, in the call to the Dex constructor that happens via BackgroundIndexRebuilder::maybeRebuild(), what path would I pass to the context provider that gets it to look at an appropriate project config file?)

you're absolutely right here. that's the main reason i don't think config mechanism actually works here. they're mostly meaningful for per-project/file configurable features. not for global ones like index.

> Only the Dex logic is conditioned on the option, not the binding of the outgoingCalls message. This is because the binding of the message happens fairly early, while we're processing the initialize request. At this point we haven't loaded config files yet, and if we try, that triggers a publishDiagnostics message for the config file, which confuses clients who are expecting an initialize response.

again this is resulting from my poor choice of words, if we treat this as part of option struct, we'll have the value at startup time and can decide what to do.

File: clang-tools-extra/clangd/index/dex/Dex.cpp [hunk 2]
```diff
@@ -149,15 +150,18 @@ void Dex::buildIndex() {
...
-    for (const auto &R : RefList)
-      if ((R.Kind & ContainedRefsRequest::SupportedRefKinds) !=
-          RefKind::Unknown)
-        RevRefs.emplace_back(R, ID);
-  // Sort by container ID so we can use binary search for lookup.
-  llvm::sort(RevRefs, [](const RevRef &A, const RevRef &B) {
-    return A.ref().Container < B.ref().Container;
-  });
+  if (Config::current().CallHierarchy.OutgoingCalls) {
+    vlog("WALDOWALDO Building RevRefs");
```
[the remark this answers is not in the document]
Comment (HighCommander4) [reply clang-tools-extra/clangd/index/dex/Dex.cpp]: Oops, debug statement left in accidentally :laughing:

**Comment by HighCommander4** [comment]
> I was trying to imply a knob like https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ClangdServer.h#L111. I don't think the knob needs to be exposed to end-users directly, only having programmatic access is fine initially.

Thanks, that is indeed quite a bit simpler.

I reworked the flag to be just a ClangdServer option, and plumbed it in explicitly, in https://github.com/llvm/llvm-project/pull/117673/commits/0aee7953ef1285bb327f20a01389e554fb44ffa8.

File: clang-tools-extra/clangd/index/FileIndex.h [hunk 3]
```diff
@@ -69,7 +69,7 @@ enum class DuplicateHandling {
 /// locking when we swap or obtain references to snapshots.
 class FileSymbols {
 public:
-  FileSymbols(IndexContents IdxContents);
+  FileSymbols(IndexContents IdxContents, bool SupportContainedRefs = true);
```
Comment (kadircet) [inline clang-tools-extra/clangd/index/FileIndex.h:72]: can we avoid the default param here?

File: clang-tools-extra/clangd/index/Serialization.h [hunk 4]
```diff
@@ -83,7 +83,8 @@ std::string toYAML(const Ref &);
 // Build an in-memory static index from an index file.
 // The size should be relatively small, so data can be managed in memory.
 std::unique_ptr<SymbolIndex> loadIndex(llvm::StringRef Filename,
-                                       SymbolOrigin Origin, bool UseDex = true);
+                                       SymbolOrigin Origin, bool UseDex = true,
+                                       bool SupportContainedRefs = true);
```
Comment (kadircet) [inline clang-tools-extra/clangd/index/Serialization.h:87]: avoiding the default here as well

File: clang-tools-extra/clangd/index/dex/Dex.h [hunk 5]
```diff
@@ -64,16 +65,18 @@ class Dex : public SymbolIndex {
...
+            std::forward<Payload>(BackingData), BackingDataSize,
+            SupportContainedRefs) {
     this->Files = std::forward<FileRange>(Files);
     this->IdxContents = IdxContents;
   }
 
   /// Builds an index from slabs. The index takes ownership of the slab.
-  static std::unique_ptr<SymbolIndex> build(SymbolSlab, RefSlab, RelationSlab);
+  static std::unique_ptr<SymbolIndex> build(SymbolSlab, RefSlab, RelationSlab,
+                                            bool SupportContainedRefs = true);
```
Comment (kadircet) [inline clang-tools-extra/clangd/index/dex/Dex.h:79]: and here

File: clang-tools-extra/clangd/tool/Check.cpp [hunk 6]
```diff
@@ -163,7 +163,7 @@ class Checker {
   unsigned ErrCount = 0;
 
   Checker(llvm::StringRef File, const ClangdLSPServer::Options &Opts)
-      : File(File), Opts(Opts) {}
+      : File(File), Opts(Opts), Index(/*SupportContainedRefs=*/false) {}
```
Comment (kadircet) [inline clang-tools-extra/clangd/tool/Check.cpp:166]: why false here? i think it's better if we use whatever clangd-server's default is

**Review by kadircet — APPROVED** [review]
thanks, mostly LG! i think not leaving the default parameter values are important here to make sure we don't forget to pass the top-level config in any of the relevant places (both now, but also in the future). i know it probably comes with a bunch of changes to tests though :/ (i wish clangd was better at updating signatures in such cases!)

File: clang-tools-extra/clangd/tool/Check.cpp (hunk 6 above)
Comment (HighCommander4) [reply clang-tools-extra/clangd/tool/Check.cpp]: My original thinking was that `--check` does not perform any outgoing calls requests. But thinking more about it, a common use of `--check` is to reproduce crashes, so the more codepaths it executes (including the one behind this flag) the better in that regard.

**Comment by HighCommander4** [comment]
Updated to remove the default arguments.

Thanks for the reviews!

File: clang-tools-extra/clangd/index/remote/Index.proto [hunk 7]
```diff
@@ -131,3 +131,21 @@ message Relation {
   optional string subject_id = 1;
   optional Symbol object = 2;
 }
+
+message ContainedRefsRequest {
+  required string id = 1;
```
Comment (kadircet) [inline clang-tools-extra/clangd/index/remote/Index.proto:136]: sorry I missed the `required` bits here and below. it's [discouraged](https://protobuf.dev/programming-guides/dos-donts/#add-required) and we're having troubles consuming this internally now.

there's no particular reason for these fields to be `required`, we're already taking care of error-handling during marshalling.

my colleague is going to revert this, I am happy to land it tomorrow with relevant changes. sorry for the second back and forth @HighCommander4  :/

**Comment by joaosaffran** [comment]
@HighCommander4 it appears that you forgot to update a [usage](https://github.com/llvm/llvm-project/blob/c3d15188cfe243900895a4f2c4f36b84e14928b7/clang-tools-extra/clangd/benchmarks/IndexBenchmark.cpp#L25-L33) of the changed loadIndex function.  This usage may be dead-coded due to the likely missing definition of a [BENCHMARK macro](https://github.com/llvm/llvm-project/blob/c3d15188cfe243900895a4f2c4f36b84e14928b7/clang-tools-extra/clangd/benchmarks/IndexBenchmark.cpp#L76-L98). Such usage broke one of HLSL build pipelines. Commenting, just for you to be aware of this issue when fixing it in the future.

File: clang-tools-extra/clangd/index/remote/Index.proto (hunk 7 above)
Comment (HighCommander4) [reply clang-tools-extra/clangd/index/remote/Index.proto:136]: No worries, please go ahead

**Comment by HighCommander4** [comment]
> relanded with benchmark fix in [61fe67a](https://github.com/llvm/llvm-project/commit/61fe67a4017375fd675f75652e857e837f77fa51)
> 
> changed required -> optional in [c7ef0ac](https://github.com/llvm/llvm-project/commit/c7ef0ac9fd28cb55b8c7c91a890b365cc688f9a9)

Thanks!

**Comment by pidgeon777** [comment]
Hello everyone, I'm seeking clarification regarding the implementation of Outgoing Calls support.

The LSP protocol currently allows to implement two display modes for visualizing outgoing function calls:

```c
#include <stdio.h>

void function1(void) {
    printf("Executing function 1\n");
}

void function2(void) {
    printf("Executing function 2\n");
}

void function3(void) {
    printf("Executing function 3\n");
}

int main(void) {
    printf("Starting main function\n");
    function1();
    function2();
    function3();
    return 0;
}
```

For example, when querying all functions called from `main`, results could be presented in two ways:

1. Display all the called function occurrences in `main`:
   - function1()
   - function2()
   - function2()
   - function3()

2. Display the definitions of the functions called in `main`:
   - function1()
   - function2()
   - function3()

According to one of the developers, this feature should provide all the necessary data to allow the LSP client to determine the presentation method (approach 1 or 2). I previously asked about this in another thread (which I cannot locate now), and I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.

**Comment by HighCommander4** [comment]
> I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.

Yes, the returned data is sufficient for clients to implement either of the above presentations.

**Comment by pidgeon777** [comment]
> > I would appreciate confirmation that this will indeed be possible with the upcoming **clangd** changes.
> 
> Yes, the returned data is sufficient for clients to implement either of the above presentations.

This is great news, thank you!

**Comment by HighCommander4** [comment]
> Any chance this is backportable to the 19.x series? Or is it too dependent on changes since then?
> 
> Or even better 18.x? (since a lot of downstream tools are dependent on 18.x and will be a while before they bump to 19)

If you're asking about the change being backported to appear in an official 19.x release, then I believe it would not be eligible for that; the release branch only accepts backports of regression and other serious bug fixes, not new features.

If you're asking whether the patch would apply cleanly to the 19.x release branch, for the purpose of applying it to some downstream repository that's based on llvm 19.x: while I haven't tried, I expect it should be possible to rebase the patch with fairly minimal changes. This is probably true of 18.x as well. If you try this and have a question about a specific conflict, I'm happy to provide suggestions.

**Comment by pidgeon777** [comment]
While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?

**Comment by HighCommander4** [comment]
> While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?

Yep

**Comment by HighCommander4** [comment]
> > @kadircet perhaps you might be able to pick up this review?
> > Or, in the absence of a full review, your opinion on the directional question in [this comment](https://reviews.llvm.org/D93829#4654786) would be appreciated as well:
> > > how would you feel about proceeding with the patch in its current state, with the memory usage increase brought down from 8.2% to 2.5% thanks to the combination of the simple lookup optimization + RefKind filtering, and leaving the "deep lookup optimization" to be explored in a future change?
> 
> I'd definitely prefer the one we discussed in the original review, but I don't think perfect needs to be enemy of the good, we can surely optimize data structures here going forward if needed.

I filed https://github.com/clangd/clangd/issues/2264 as a follow-up to track implementation of the "deep lookup optimization".

**Comment by pidgeon777** [comment]
> > While I don't have direct experience with LLVM releases, I suppose the Outgoing Calls feature will be part of version 20?
> 
> Yep

Does anyone know the estimated release date for version 20?

**Comment by HighCommander4** [comment]
> Does anyone know the estimated release date for version 20?

Based on https://llvm.org/docs/HowToReleaseLLVM.html#annual-release-schedule, approximately March 6.

**Pull request merged**